pkg/apis/kops/cluster.go

```diff
@@ -128,6 +128,7 @@ type ClusterSpec struct {
 	// Component configurations
 	Docker                         *DockerConfig                 `json:"docker,omitempty"`
 	KubeDNS                        *KubeDNSConfig                `json:"kubeDNS,omitempty"`
+	CoreDNS                        *CoreDNSConfig                `json:"coreDNS,omitempty"`
```

Comment:
Another option is to have:

```
kubeDNS:
  domain: ...
  serverIP: ...
  clusterIP: ...
  provider: coredns
```

What I like about that is that we're not duplicating the field validation etc.  What I dislike is that the field is called `kubeDNS`, and presumably the options for coreDNS will start to diverge.

We can fix the field name in the next API version, but the options problem is more fundamental, and does support the approach you've taken here.

Reply:
Yes ... much better. The name ugh. Beta APIs!!!!

Reply:
@johnbelamaric I am very confused, do we have both installed on the same cluster?

Reply:
Sorry, let me clarify. No, we won't have both running. However, during an upgrade, we want to bring CoreDNS up, and then take Kube-DNS down - without a loss of service. We can't create Service named `coredns` with the same clusterIP as the `kube-dns` Service, because we would have to delete the `kube-dns` service first. That would lead to a (brief) DNS outage.

In the `kubeadm` discussion on upgrade, we decided that the name of the Service would remain `kube-dns` - to imply simply that it is the Kubernetes in-cluster DNS, not that it is served by the code currently known as `kube-dns`.

Leaving this name the same allows us to just point the existing Service at the new, CoreDNS deployment, preventing any loss of DNS service during the upgrade process.

Since we needed to do this for upgrade, we left the name the same for new installs too. So, whether your cluster DNS is actually handled by kube-dns or CoreDNS, the name of the Kubernetes Service is `kube-dns`.